tor: enable PoW defenses for automatically created hidden services

vasild opened the pull request:
Enable [PoW defenses](https://tpo.pages.torproject.net/onion-services/ecosystem/technology/security/pow/) for hidden services that we create via Tor Control using the [`ADD_ONION` command](https://spec.torproject.org/control-spec/commands.html#add_onion).

The ability to do that has been added in [tor-0.4.9.2-alpha](https://gitlab.torproject.org/tpo/core/tor/-/commit/02c18044464bfe45f168b55297a785244094cfd5). Previous versions return a syntax error to the `ADD_ONION` command with `PoWDefensesEnabled=1`, so the approach here is to try with PoW and if we get syntax error, then retry without PoW.

Also update `doc/tor.md` with a hint on enabling PoW on manually configured Tor hidden services.

vasild committed:
tor: enable PoW defenses for automatically created hidden services

Enable PoW defenses [1] for hidden services that we create via
Tor Control using the `ADD_ONION` command [2].

The ability to do that has been added in tor-0.4.9.2-alpha [3]. Previous
versions return a syntax error to the `ADD_ONION` command with
`PoWDefensesEnabled=1`, so the approach here is to try with PoW and if
we get syntax error, then retry without PoW.

[1] https://tpo.pages.torproject.net/onion-services/ecosystem/technology/security/pow/
[2] https://spec.torproject.org/control-spec/commands.html#add_onion
[3] https://gitlab.torproject.org/tpo/core/tor/-/commit/02c18044464bfe45f168b55297a785244094cfd5

vasild committed:
doc: add a hint to enable PoW defenses to manual hidden services

vasild committed:
doc: add release notes for Tor PoW defenses

willcl-ark commented:
> Should we then also add PoW to the connections that we make to other nodes running behind hidden services?

Reading the linked FAQ, the feature still supports "older clients" (which don't have PoW defence capability), but they may take a lower priority when a service considers itself under DoS. So no PoW is _required_ on the client side.

When the client-side tor is new-enough, my understanding is that the puzzle-solving is automatically handled by Tor, and doesn't need client-side changes to the connection code, as it happens during the introduction. But I am not 100% certain.

fanquake commented:
What is the status of this?

willcl-ark commented on src/test/fuzz/torcontrol.cpp:
I think `tor_control_reply_code = 512` could be added here, to hit this expected path more frequently?

vasild replied:
Right, good catch! Done and also moved the `TOR_REPLY_*` constants to `torcontrol.h` so this fuzz test can use them instead of hardcoding numbers. Thanks!

willcl-ark reviewed (commented):
Approach and lightly-tested ACK on PoW-enabled Tor (version 0.4.8.22).

```
2026-02-16T21:44:26Z [tor] Get SOCKS port command yielded 127.0.0.1:9050
2026-02-16T21:44:26Z [tor] Configuring onion proxy for 127.0.0.1:9050
2026-02-16T21:44:26Z [tor] ADD_ONION failed with PoW defenses, retrying without
2026-02-16T21:44:26Z [tor] ADD_ONION successful (PoW defenses disabled)
```

Seems very reasonable to me to implement this configuration in order that we fare better during DoS attacks on the Tor network. Noting for others that as far as I read, due to the dynamic difficulty the (tor) PoW is ~ free on idle, and the cost only applies during an attack, which seems nice.

I quite like the pragmatism of "detecting" the tor version via the failure mode. It appears that the 512 failure happens early on the Tor side, before any service is created. So failing and retrying is "clean".

> What is the status of this?

Looks like it needs a pretty trivial rebase at the moment is all.

vasild committed:
tor, fuzz: reuse constants instead of duplicating

`src/torcontrol.cpp` used to define some constants that are used
explicitly in `src/torcontrol.cpp` and implicitly in
`src/test/fuzz/torcontrol.cpp` by duplicating their values.

Move the constants to `src/torcontrol.h` and reuse them in
`src/test/fuzz/torcontrol.cpp` to avoid duplication and magic
numbers.

willcl-ark reviewed (approved):
ACK c68e3d2c57dcab5cea22ad5986fcd2b147a7daaa

Tested with two versions of `tor`. ON 0.4.8.22 automatic PoWDefense via ONION message (correctly) does not work and falls back to without.

On 0.4.9.5 I see:

```
2026-02-19T12:04:04Z torcontrol thread start
2026-02-19T12:04:04Z [tor] Reading cached private key from /xxxxxxxx/onion_v3_private_key
2026-02-19T12:04:04Z [tor] Successfully connected!
2026-02-19T12:04:04Z [tor] Connected to Tor version 0.4.9.5
2026-02-19T12:04:04Z [tor] Supported authentication method: NULL
2026-02-19T12:04:04Z [tor] Using NULL authentication
2026-02-19T12:04:04Z [tor] Authentication successful
2026-02-19T12:04:04Z [tor] Get SOCKS port command yielded 127.0.0.1:9050
2026-02-19T12:04:04Z [tor] Configuring onion proxy for 127.0.0.1:9050
2026-02-19T12:04:04Z [tor] ADD_ONION successful (PoW defenses enabled)
```

The changes look clean and correct to me.

fjahr commented:
Concept ACK

fjahr reviewed (commented):
tACK c68e3d2c57dcab5cea22ad5986fcd2b147a7daaa

Only tested that PoW enable worked with the latest tor version (0.4.9.5), not the fallback with an older tor version. Code looks good, note to self: I will extend the test coverage in #34158 with additional coverage for this.

sedited reviewed (approved):
ACK c68e3d2c57dcab5cea22ad5986fcd2b147a7daaa
